File: OC_JS_WKWebView/OC_JS_WKWebView/LWJSWebView.m
//
//  LWJSWebView.m
//  OC_JS_WKWebView
//
//  Created by liwei on 2018/7/9.
//  Copyright © 2018年 liwei. All rights reserved.
//

#import "LWJSWebView.h"

// ================================================================================================================================================
#pragma mark - js交互网页
@interface LWJSWebView()<WKUIDelegate,WKScriptMessageHandler>

@end
// ================================================================================================================================================
#pragma mark - js交互网页
@implementation LWJSWebView

#pragma mark - 懒加载网页视图
- (WKWebView *)wkWebView {
    if (_wkWebView == nil) {
        WKWebViewConfiguration *configuration = [[WKWebViewConfiguration alloc] init];
        WKPreferences *preferences = [WKPreferences new];
        preferences.javaScriptCanOpenWindowsAutomatically = YES;
        preferences.minimumFontSize = 40.0;
        configuration.preferences = preferences;
        _wkWebView = [[WKWebView alloc] initWithFrame:CGRectZero configuration:configuration];
//        _wkWebView = [[WKWebView alloc] initWithFrame:CGRectZero];
        _wkWebView.UIDelegate = self;
    }
    return _wkWebView;
}

#pragma mark - 懒加载方法数组
- (NSMutableArray<LWJSWebViewSelectorModel *> *)funcArray {
    if (_funcArray == nil) {
        _funcArray = [NSMutableArray array];
       NSString *path = [[NSBundle mainBundle] pathForResource:@"JSSelector" ofType:@".plist"];
        NSArray *arrayDict = [NSMutableArray arrayWithContentsOfFile:path];
        for (NSDictionary *dict in arrayDict) {
            LWJSWebViewSelectorModel *model = [[LWJSWebViewSelectorModel alloc] init];
            model.selectorName = dict[@"selectorName"];
            model.needPar = dict[@"needPar"];
            [_funcArray addObject:model];
        }
    }
    return _funcArray;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self addSubview:self.wkWebView];
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.wkWebView.frame = self.bounds;
}

- (void)loadRequest:(NSURLRequest *)request {
    [self.wkWebView loadRequest:request];
}
- (void)loadFileURL:(NSURL *)URL allowingReadAccessToURL:(NSURL *)readAccessURL  {
    [self.wkWebView loadFileURL:URL allowingReadAccessToURL:readAccessURL];
}
- (void)loadHTMLString:(NSString *)string baseURL:(nullable NSURL *)baseURL {
    [self.wkWebView loadHTMLString:string baseURL:baseURL];
}

#pragma mark 添加handlers
- (void)addScriptMessageHandler {
    [self removeScriptMessageHandler];
    [self.funcArray enumerateObjectsUsingBlock:^(LWJSWebViewSelectorModel * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        [self.wkWebView.configuration.userContentController addScriptMessageHandler:self name:obj.selectorName];
    }];
}
#pragma mark 移除handlers
- (void)removeScriptMessageHandler {
    [self.funcArray enumerateObjectsUsingBlock:^(LWJSWebViewSelectorModel * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        [self.wkWebView.configuration.userContentController removeScriptMessageHandlerForName:obj.selectorName];
    }];
}
#pragma mark  WKScriptMessageHandler
- (void)userContentController:(nonnull WKUserContentController *)userContentController didReceiveScriptMessage:(nonnull WKScriptMessage *)message {
    //    message.body  --  Allowed types are NSNumber, NSString, NSDate, NSArray,NSDictionary, and NSNull.
       NSLog(@"body:%@",message.body);
        LWJSWebViewSelectorModel *selectorModel = nil;
        for (LWJSWebViewSelectorModel *model in self.funcArray) {
            if ([model.selectorName isEqualToString:message.name]) {
                selectorModel = model;
                break;
            }
        }
    if (selectorModel == nil) {
        return;
    }
    SEL selector = NSSelectorFromString(selectorModel.selectorName);
    if (selectorModel.needPar == YES) {
        selector = NSSelectorFromString([NSString stringWithFormat:@"%@:",selectorModel.selectorName]);
    }
    if ([self respondsToSelector:selector]) {
# pragma clang diagnostic push
# pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        if (selectorModel.needPar == YES) {
            [self performSelector:selector withObject:message.body];
        }
        else {
            [self performSelector:selector];
        }
# pragma clang diagnostic pop
    }
}
#pragma mark - 调用方法
- (void)OCtoJS:(id)par {
    
}
- (void)JStoOC:(id)par {
    // 将结果返回给js
    NSString *jsStr = [NSString stringWithFormat:@"JStoOCResult('%@')",@"回调成功"];
    [self.wkWebView evaluateJavaScript:jsStr completionHandler:^(id _Nullable result, NSError * _Nullable error) {
        NSLog(@"%@----%@",result, error);
    }];
}
@end
// ================================================================================================================================================
#pragma mark - js交互网页方法模型
@implementation LWJSWebViewSelectorModel

@end
// ================================================================================================================================================

File: OC_JS_WKWebView/OC_JS_WKWebView/LWJSWebView.swift
//
//  LWJSWebView.swift
//  swift_JS_WKWebView
//
//  Created by liwei on 2018/7/10.
//  Copyright © 2018年 liwei. All rights reserved.
//

import UIKit
import WebKit
// ================================================================================================================================
// MARK: - js交互网页
class LWJSWebView: UIView {

    /// 方法数组
    var funcArray = [LWJSWebViewSelectorModel]()
    /// 网页视图
    lazy var wkWebView: WKWebView = {
        let conf = WKWebViewConfiguration.init()
        let pref = WKPreferences.init()
        pref.javaScriptCanOpenWindowsAutomatically = true
        conf.preferences = pref
        let webView = WKWebView.init(frame: CGRect.zero, configuration: conf)
        webView.uiDelegate = self
        return webView
    }()
    override init(frame: CGRect) {
        super.init(frame: frame)
        let path = Bundle.main.path(forResource: "JSSelector", ofType: ".plist")
        let arrayDict = NSArray.init(contentsOfFile: path!)
        let count : Int = (arrayDict?.count)!
        for index in 0..<count {
            let dict = arrayDict![index] as! NSDictionary
            let model = LWJSWebViewSelectorModel.init()
            model.selectorName = dict["selectorName"] as! String
            model.needPar = dict["needPar"] as! Bool
            funcArray.append(model)
        }
        self.addSubview(self.wkWebView)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    override func layoutSubviews() {
        super.layoutSubviews()
        self.wkWebView.frame = self.bounds
    }
    
    ///  添加handlers
    func addScriptMessageHandler() {
        removeScriptMessageHandler()
        for (_,value) in self.funcArray.enumerated() {
            self.wkWebView.configuration.userContentController.add(self, name: value.selectorName)
        }
    }
    ///   addScriptMessageHandler 很容易导致循环引用控制器 强引用了WKWebView,WKWebView copy(强引用了）configuration， configuration copy （强引用了）userContentControlleruserContentController 强引用了 self （控制器）这里要记得移除handlersviewWillDisappear:
    func removeScriptMessageHandler() {
        for (_,value) in self.funcArray.enumerated() {
            self.wkWebView.configuration.userContentController.removeScriptMessageHandler(forName: value.selectorName)
        }
    }
    
    
    
    @objc  func OCtoJS() {
        print("OCtoJS")
    }
    
    
    
}
// ================================================================================================================================
// MARK: - js交互网页代理
extension LWJSWebView : WKUIDelegate,WKScriptMessageHandler {
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        print("body:\(message.body)")
        var model : LWJSWebViewSelectorModel? = nil
        for (_,value) in self.funcArray.enumerated() {
            if value.selectorName == message.name {
                model = value
                break
            }
        }
        if model == nil {
            return
        }
        var selector = Selector.init((model?.selectorName)!)
        if ((model?.needPar = true) != nil) {
            selector = Selector.init(String(describing: (model?.selectorName)! + ":"))
        }
        if responds(to: selector) {
            if ((model?.needPar = true) != nil) {
                perform(selector, with: message.body)
            }
            else {
                perform(selector)
            }
        }
    }
}
// ================================================================================================================================
// MARK: - js交互网页模型
class LWJSWebViewSelectorModel : NSObject {
    
    /// 方法名
    var selectorName = ""
    /// 是否需要参数
    var needPar = false
    
    override init() {
        
    }
    
}
// ================================================================================================================================
